tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc

```diff
@@ -117,51 +345,246 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+      ctx->SetStatus(status);
+      return;
     }
-
-    OP_REQUIRES_OK(context,
-                   context->allocate_output(i, output_shape, &output_tensor));
     auto dtype = trt_engine_ptr_->getBindingDataType(binding_index);
     switch (dtype) {
       case nvinfer1::DataType::kFLOAT:
         buffers[binding_index] =
             reinterpret_cast<void*>(output_tensor->flat<float>().data());
         break;
       case nvinfer1::DataType::kHALF:
-        LOG(FATAL) << "half size is not supported yet!";
+        LOG(ERROR) << "half size is not supported yet!";
+        ctx->SetStatus(tensorflow::errors::InvalidArgument(
+            "Half outputs are not supported!"));
+        return;
         break;
       case nvinfer1::DataType::kINT8:
-        LOG(FATAL) << "int8 is not supported yet!";
+        LOG(ERROR) << "int8 is not supported yet!";
+        ctx->SetStatus(tensorflow::errors::InvalidArgument(
+            "INT8 outputs are not supported!"));
+        return;
         break;
       default:
-        LOG(FATAL) << "Unknown data type: " << int(dtype);
+        LOG(ERROR) << "Unknown TRT data type: " << int(dtype);
+        ctx->SetStatus(tensorflow::errors::InvalidArgument(
+            "Unsupported output data type! " + int(dtype)));
+        return;
         break;
     }
   }
   // copied from cuda_kernel_helper since it seems only valid in *.cu.cc files
   const cudaStream_t* stream = CHECK_NOTNULL(
-      reinterpret_cast<const cudaStream_t*>(context->op_device_context()
+      reinterpret_cast<const cudaStream_t*>(ctx->op_device_context()
                                                 ->stream()
                                                 ->implementation()
                                                 ->CudaStreamMemberHack()));
 
   // TODO(jie): trt enqueue does not return error
-  auto ret = trt_execution_context_ptr_->enqueue(num_batch, &buffers[0],
-                                                 *stream, nullptr);
+  auto trt_execution_context_ptr = engine_ctx_pair.second;
+  auto ret = trt_execution_context_ptr->enqueue(num_batch, &buffers[0], *stream,
+                                                nullptr);
   VLOG(2) << "enqueue returns: " << ret;
   // sync should be done by TF.
 }
+
 TRTEngineOp::~TRTEngineOp() {
   // Order matters!
-  trt_execution_context_ptr_.reset();
-  trt_engine_ptr_.reset();
-  allocator_.reset();
+  for (auto eng : engine_map_) {
+    eng.second.first.reset();
+    eng.second.second.reset();
+  }
+  for (auto alloc : allocators_) alloc.second.reset();
+}
+
+TRTEngineOp::EngineCtxPair TRTEngineOp::GetEngine(int batch_size,
+                                                  OpKernelContext* ctx,
+                                                  bool ignore_dim_change) {
+  // TODO(sami): This method needs to be re-written to use resource manager and
+  // with LRU mechanism option.
+  tensorflow::mutex_lock lock(engine_mutex_);
+  if (static_engine_) {
```

Comment:
No, Static engine is for the cases where everything is fixed and can be determined offline. For such networks there is no point of constructing the engine on the fly. These mostly for final deployment cases.

Reply:
There are cases where initialization of the engine is taking in the order or tens of minutes. It is not acceptable to ask every instantiation of an inference network wait for tens of minutes before it can process any input. For such networks, portability is not an issue, there is only one target device. Portability and maintenance argument applies to anything in TensorFlow. If there are backward incompatible changes, user needs to update the graph. Also the flag is there for a reason. There are users which want dynamic construction and there are users which want static construction. Each has pros and cons and users are free to choose whichever is suitable for them. We want users to be able to choose. Static engine is backwards compatible behavior. We may switch the default to dynamic engine but I would prefer doing that gradually, after users have a chance to try and see which one is most suitable for them. In short static construction is a needed feature.